Add support for Unity NativeArray as vector type

joncham opened the pull request:
Implementation for issue https://github.com/jamescourtney/FlatSharp/issues/303

Initial work to support Unity's NativeArray as a supported vector type.

### Highlights:

- New C# project `FlatSharp.Compiler.UnityStub` (naming?) added to provide stub/polyfill APIs needed for schema compilation as well as test compilation.
- The C# file `src/FlatSharp.Compiler.UnityStub/UnityExtensions.cs` provides the glue to Unity APIs.
   - When built via the csproj (`FLATSHARP_UNITY_POLYFILLS` define) the `NativeArray` type is stubbed and extension methods for reading/writing are empty, allowing the usage of types and API surfaces but no functionality.
    - When this file is included in a Unity project, `NativeArray` is provided by Unity assemblies and the reading/writing methods are implemented against further Unity APIs.
- Note that `FlatSharp.Compiler.UnityStub` assembly will be deployed with the compiler, but should never be used at runtime in a Unity context

### Open Questions
1. I initially had the Unity APIs withing the FlatSharp.Compiler assembly itself. This worked until I started writing tests, as they APIs would either need to be public or the internals would need exposed to tests. Is the current "stub" assembly approach acceptable?
2. The `NativeArray` is constructed from the buffer/memory passed into `ReadNativeArrayBlock`. **A copy is not performed.** Is it safe to assume this memory is native or pinned managed memory that will not be moved?
3. I added test for simple usage of NativeArray which just validates the compiler generates proper code. I can add tests for actual functionality by further stubbing out Unity APIs. Any input on this?

Thanks!

jamescourtney commented on src/FlatSharp.Compiler.UnityStub/NativeArray.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace Unity.Collections;
+
+public struct NativeArray<T> : IDisposable, IEnumerable<T>, IEquatable<NativeArray<T>> where T : struct
```
Is this intended to be immutable? I was thinking it would be nice to have at least a modest implementation here so we could write some functional-ish tests.

joncham replied:
I can look at filling this in, although it won't be possible to match the exact implementation from Unity as it assumes it's running within a Unity Editor/Player environment.

joncham replied:
Not immutable, but I have added a basic implementation and a number of other tests.

jamescourtney commented on src/FlatSharp.Compiler.UnityStub/UnityExtensions.cs:
```diff
@@ -0,0 +1,89 @@
+            // The assumption is this is native memory or pinned beyond 'Pin' call below. 
+            using (var pin = mem.Pin())
+            {
+                var na = NativeArrayUnsafeUtility.ConvertExistingDataToNativeArray<TElement>(pin.Pointer, mem.Length, Allocator.Invalid);
```
Is `NativeArrayUnsafeUtility` defined somewhere else and the #ifdef's only apply when building not-for-Flatsharp but in some other repo?

Assuming this does a copy, right? One of the things I've tried to avoid in FlatSharp is that when using Lazy deserialization, we try to avoid doing Greedy things like array copies. I've had to part with that restriction recently, but it still makes me grimace a little bit.

I guess with Native memory, you don't have a GC overhead or anything like that, so it's not too bad if the array is small-ish.

vvuk replied:
Yeah, it's defined as part of Unity's APIs.  We should be able to provide a stub.  It converts an existing bare pointer/length to a NativeArray; doesn't do a copy.

joncham replied:
Stub has been added and is discussed in a different issue.

jamescourtney commented on src/FlatSharp.Compiler.UnityStub/UnityExtensions.cs:
```diff
@@ -0,0 +1,89 @@
+            fixed (void* p = span.Slice(vectorStartOffset + sizeof(uint)))
+            {
+                UnsafeUtility.MemCpy(p, buffer.GetUnsafeReadOnlyPtr(), buffer.Length * UnsafeUtility.SizeOf<TElement>());
+            }
```
Would some application of `MemoryMarshal.Cast` and `Span.CopyTo` work here?

vvuk replied:
Hmm.  I didn't realize Span<T> had a (pointer, length) constructor.  There's a chance we don't have full support for Span/MemoryMarshal but if we do, yeah, seems like this could be something like `var src = new Span<TElement>(buffer.GetUnsafeReadOnlyPtr(), buffer.Length); src.CopyTo(span.Slice(...));`

joncham replied:
I have migrated to standard .NET BCL APIs. I'll need to check on performance of these in Unity scenario (running on Mono & IL2CPP runtimes).

jamescourtney commented on src/FlatSharp.Compiler/FlatSharpCompiler.cs:
```diff
@@ -522,7 +526,7 @@ void OnDataReceived(object sender, DataReceivedEventArgs args)
                         InputHash = inputHash,
                         PreviousAssembly = assembly,
-                        TypeModelContainer = TypeModelContainer.CreateDefault(),
+                        TypeModelContainer = TypeModelContainer.CreateDefault().WithUnitySupport(),
```
Wondering if this should be a compiler switch like `--enable-unity-support`? I don't think I have strong feelings either way, so it's probably fine.

joncham replied:
I have added `--unity-assembly-path`, which does also require a path argument, to enable Unity support.

jamescourtney commented on src/FlatSharp.Compiler/UnityNativeArrayVectorTypeModel.cs:
```diff
@@ -0,0 +1,66 @@
+
+    public override CodeGeneratedMethod CreateCloneMethodBody(CloneCodeGenContext context)
+    {
+        string body = $"throw new NotSupportedException();";
```
this will break copy constructors that the compiler generates. Since it fails predictably, that seems not-terrible, but it's going to be a non-obvious failure. Is there a case where NativeArrays should be cloneable? Since they seem immutable it might be enough to just do the equivalent of what we do for strings and primitives where we just return the original value and the whole thing becomes a no-op.

joncham replied:
They are mutable. I am looking into handling some sort of Clone now.

jamescourtney commented on src/FlatSharp.Compiler/UnityTypeModelProvider.cs:
```diff
@@ -0,0 +1,42 @@
+        if (type.IsGenericType)
+        {
+            var genericDef = type.GetGenericTypeDefinition();
+            if (genericDef.Name == "NativeArray`1")
```
Does `genericDef == typeof(NativeArray<>)` not work here?

jamescourtney replied:
I didn't think this particular code ran in the Unity environment since this is the Flatsharp type mode and is only used for codegen?

joncham replied:
Fixed.

jamescourtney commented on src/Tests/FlatSharpCompilerTests/InvalidAttributeTests.cs:
```diff
@@ -139,6 +139,6 @@ namespace ns;
 
         var ex = Assert.Throws<InvalidFbsFileException>(() => FlatSharpCompiler.CompileAndLoadAssembly(schema, new()));
-        Assert.Contains("Unable to parse 'fs_vector' value 'banana'. Valid values are: IList, IReadOnlyList, Array, Memory, ReadOnlyMemory, IIndexedVector.", ex.Message);
+        Assert.Contains("Unable to parse 'fs_vector' value 'banana'. Valid values are: IList, IReadOnlyList, Array, Memory, ReadOnlyMemory, IIndexedVector, NativeArray.", ex.Message);
```
Reading this, I think I'd prefer the naming to be `Unity_NativeArray` or `UnityNativeArray`. I'd like to indicate that it's definitely a Unity-only thing.

joncham replied:
Renamed to `UnityNativeArray`

jamescourtney commented on src/Tests/FlatSharpCompilerTests/TableMemberTests.cs:
```diff
@@ -90,6 +91,9 @@ private void RunCompoundTestWithDefaultValue_Bool(string fbsType)
         this.RunSingleTest<T>($"{fbsType} = {randomValue.ToString(format, null).ToLowerInvariant()}", hasDefaultValue: true, expectedDefaultValue: randomValue);
         this.RunCompoundTest<T>(fbsType);
+#if NETSTANDARD2_1 || NET5_0_OR_GREATER
+        this.RunSingleTest<NativeArray<T>?>($"[{fbsType}]  ({MetadataKeys.VectorKind}:\"NativeArray\")");
```
So, the only test is just that it builds? If we were to supply some mock implementations could we actually test it?

joncham replied:
Stubbed and tests added.

jamescourtney reviewed (changes requested):
Overall, this looks like you're on the right track. Thanks for making the code look like the rest of the code in this repo (writing code to generate code is often awkward -- let me know if you know of any better techniques!). I do have some questions about where the implementations of the other parts of the `#ifdef` declarations live, and I'd like to see more robust testing, if that's something that's practical.

joncham commented on src/FlatSharp.Compiler.UnityStub/FlatSharp.Compiler.UnityStub.csproj:
```diff
@@ -0,0 +1,24 @@
+  <Import Project="..\common.props" />
+  
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;netstandard2.1;netcoreapp3.1;net5.0;net6.0</TargetFrameworks>
```
Fixed.

joncham commented:
I pushed some changes that address many of the issues raised and with some other differences. I can add more details later, but quickly:
1. New command line option `--unity-assembly-path` which a) enables Unity support b) takes a path to an assembly providing Unity APIs
2. Moved some types from compiler to runtime assembly. This allows Unity support to work in runtime compilation mode (not just AOT)
3. Further stubbed out NativeArray. This implementation is quite different than Unity but functional.
4. Implemented shared code path for NativeArray (de)serialization. This makes a copy of the underlying buffer data.
5. Added a bunch of tests.

Next, I'll look into proper lazy deserialization/pinned support for NativeArray deserialization (avoiding the copy).

joncham commented on src/FlatSharp.Runtime/IO/MemoryInputBuffer.cs:
```diff
@@ -25,12 +25,14 @@ public struct MemoryInputBuffer : IInputBuffer
     private readonly MemoryPointer pointer;
 
-    public MemoryInputBuffer(Memory<byte> memory)
+    public MemoryInputBuffer(Memory<byte> memory, bool pinned = false)
```
@jamescourtney I added `IInputBuffer.IsPinned` and added parameter with default value to constructor of `MemoryInputBuffer` and `ReadOnlyMemoryInputBuffer` as I that was practical approach. I am fine adding a new buffer type and/or removing the `IsPinned` and keying off the type. Let me know what you think.

jamescourtney replied:
I think IsPinned is a great compromise.

jamescourtney commented on src/FlatSharp/TypeModel/UnityTypeModelProvider.cs:
```diff
@@ -0,0 +1,38 @@
+﻿/*
+ * Copyright 2022 Unity Technologies
```
I opened another comment about this in the past, but just so you know, I'm going to change all copyright headers in the entire repo to read "Copyright 2022 by the FlatSharp contributors" after you commit your changes. I'm happy to add Unity Technologies to an official contributor list as well -- please let me know if you have a preference.

jamescourtney replied:
https://opensource.stackexchange.com/questions/7300/copyright-notice-in-the-file-header-apache-v2-license

jamescourtney commented on src/FlatSharp/TypeModel/Vectors/UnityNativeArrayVectorTypeModel.cs:
```diff
@@ -0,0 +1,57 @@
+            context.AllFieldContexts,
+            context.Options);
+
+        var lazySuffix = context.Options.Lazy ? "Lazy" : "";
```
Sorry -- it looks like you got caught up in a merge conflict here. FlatSharp 7 is going to have a different characterstic, which is to generate all serializers for all types. The broader goal is to reduce code duplication for better cache hits and branch predictor utilization for cases when two parent tables both reference the same inner table. In that case, FlatSharp 6 will generate an entirely different serializer "graph" for each parent table, so all shared dependencies have duplicate code.

jamescourtney commented on src/FlatSharp/TypeModel/Vectors/UnityNativeArrayVectorTypeModel.cs:
```diff
@@ -0,0 +1,57 @@
+    public override CodeGeneratedMethod CreateSerializeMethodBody(SerializationCodeGenContext context)
+    {
+        var writeNativeArray = $"{context.SpanWriterVariableName}.WriteNativeArrayBlock({context.SpanVariableName}, {context.ValueVariableName}, {context.OffsetVariableName}, " +
+                               $"{ItemTypeModel.PhysicalLayout[0].Alignment}, {this.PaddedMemberInlineSize}, {context.SerializationContextVariableName});";
```
Let's add a validation that PhyiscalLayout is of length 1 for these. PhysicalLayout really means "vtable layout". Unions in Flatbuffers are done... strangely and are the only types that have a physical layout of length, so this really means that you can't store unions in a NativeArray. Since FlatSharp unions are structs, someone could actually get the wrong idea here :)

jamescourtney replied:
Sorry, to expand on this, validations are usually done in the `Initialize` method, and failures usually throw a `InvalidFlatBufferDefinitionException` with some human-readable message.

Validations I think you need here are...:
- Value type (unless the generic definition prevents this).
- Not a union (physical layout length == 1)
- Item is a valid vector member (ItemTypeModel.IsValidVectorMember)

jamescourtney commented on src/FlatSharp.Compiler/SchemaModel/PropertyFieldModel.cs:
```diff
@@ -317,6 +317,7 @@ private string GetVectorTypeName(string innerType, string? keyType)
             VectorType.Memory => $"Memory<{innerType}>",
             VectorType.ReadOnlyMemory => $"ReadOnlyMemory<{innerType}>",
             VectorType.IIndexedVector => $"IIndexedVector<{keyType ?? "string"}, {innerType}>",
+            VectorType.UnityNativeArray => $"Unity.Collections.NativeArray<{innerType}>",
```
One thing I think will be a problem is that there is no check here to enforce that the type of inner value is actually a value type (ref struct). My suggestion is twofold:

- Relax the generic type restrictions in the polyfill to allow any type (so you don't get compile errors using a reference type)
- Add validation to the new type model to ensure that the inner member is a...
   - Value type / primitive
   - Value struct
   
But I think we want to exclude tables here as well as strings and unions.

jamescourtney commented on src/FlatSharp.UnityPolyfills/NativeArray.cs:
```diff
@@ -0,0 +1,150 @@
+    {
+        internal Memory<T> m_Buffer;
+
+        internal void* m_Data;
```
There isn't a need to use a pointer here or anything like that. I'd be content with an inner array. I just wanted _some_ way to test this code -- I don't want to mock Unity too closely, and am comfortable assuming that anything matching the correct API surface will work in Unity.

jamescourtney reviewed (changes requested):
Thanks for the changes here! I have some more comments, but they are a bit lower level than before, which is usually a good sign.

Apologies for two different reviews. I did the first one and then realized I had more comments, which I started adding and then realized it might be easier for you if I had unified everything together.

I mentioned this in one comment, but a lot of times I just ask questions in PRs for the sake of facilitating discussion; if you feel strongly, please disagree with me and we can chat about it!

joncham commented on src/FlatSharp.UnityPolyfills/UnityExtensions.cs:
```diff
@@ -0,0 +1,96 @@
+{
+    public static class UnityExtensions
+    {
+        public static NativeArray<TElement> ReadNativeArrayBlock<TElement, TBuffer>(this TBuffer buffer, int uoffset)
```
> * Will generated code need to depend on this Polyfill assembly like it does FlatSharp.Runtime?

It will depend on `UnityExtensions.cs` only. The `NativeArray.cs` file will be fulfilled by assemblies present in the Unity environment.

> But in that case, we won't care which NativeArray is referenced, only that the type exists.
This should be the case today. Tests run with the `NativeArray` stubbed here. Unity uses it's own implementation.

> Would it perhaps be better to emit this from the Type Model codegen instead of a method here?
I have debated this, but I still need to resolve a `NativeArray` type. Since I already will require an external assembly reference for the `NativeArray` type, the hand written C# has not been too burdensome. Also, it's easier to reason about and iterate on as a static file ATM.

jamescourtney replied:
Since UnityExtensions is in a different DLL, then we'll definitely need to push that as a separate package, correct?

joncham commented on src/FlatSharp.UnityPolyfills/NativeArray.cs:
```diff
@@ -0,0 +1,150 @@
+﻿/*
+ * Copyright 2022 Unity Technologies
```
Sounds good. I am chasing some things regarding this on my side, but don't see any issue with the proposed changes.

joncham commented:
@jamescourtney is it worth splitting this PR up a bit? For example the `IsPinned` property on the buffer interface/implementations?

jamescourtney commented:
> @jamescourtney is it worth splitting this PR up a bit? For example the `IsPinned` property on the buffer interface/implementations?

I'm comfortable with the size of it, but I'll be happy to go whatever way you want. I'm also happy to check in and iterate as long as we have items to track various discussion points.

joncham commented:
Okay, with 327fe1feda803e96904f29d5174cb79bb9282e74 I have:
1. Added helper methods to `InputBufferExtensions` and `SpanWriterExtensions` to read/write from a Span
2. Removed Unity `src/FlatSharp.UnityPolyfills/UnityExtensions.cs`
3. Codegen'd the body of the NativeArray serialize/parse in UnityNativeArrayVectorTypeModel, using only Unity public APIs and the new helpers added in 1.

This means `src/FlatSharp.UnityPolyfills` is solely for testing purposes and should only contain stubs for existing Unity public APIs. It should not need included in a new or existing NuGet package. I'll rename it in a subsequent PR. Should it move until the Tests directory?

It also means that no extra source file is needed for "deployment" into Unity anymore implementing the read/write methods.

joncham commented:
And thanks for the help and feedback @jamescourtney , I think we are iterating to a good solution.

jamescourtney commented on src/FlatSharp.Runtime/IO/SpanWriterExtensions.cs:
```diff
@@ -36,6 +38,27 @@ public static class SpanWriterExtensions
+    {
+        int numberOfItems = buffer.Length;
+        int vectorStartOffset = ctx.AllocateVector(
+            itemAlignment: Unsafe.SizeOf<TElement>(),
```
In FlatBuffers, structs are aligned to according to the max alignment of all their members. So a struct with an int64 member would have 8-byte alignment, etc. This is an artifact of the time when FlatBuffers was designed; some architectures at the time didn't support unaligned reads, so care was taken to align all the data. This turns out to be not a good choice for size on the wire, but the format is the format.

I think the case where a memory copy will be enough to properly serialize a struct is actually pretty narrow: There is no padding between members (that is -- `StructSize % Alignment == 0`). An example would be the typical `Vector3` struct with 3 float32's inside.

However, there are other cases where a memory copy won't be enough:

- Big endian machines. FlatBuffers encodes all data as little endian. FlatSharp at least has code sprinkled throughout trying to be thoughtful about BE architectures. Not that I have a BE machine to test it on, of course, But it's there. And it might even work :)

- Cases where consecutive structs need padding between items (ie, not aligned according to the above).

The fallback for both of these cases would be to use a loop and serialize/parse "normally". You can take a look at what I do for value-type structs [here](https://github.com/jamescourtney/FlatSharp/blob/2ec1312212655f5bf05e9678f3ec205aad3a410e/src/FlatSharp/TypeModel/ValueStructTypeModel.cs#L171). Basically, the logic is that if memory-marshalling is allowed by the user and it's a LE machine, then do it. Otherwise, go field-by-field.

That said, if Unity is LE only/mostly and you think it would be better to raise the error about poor alignment rather than have people get unexpectedly bad performance, then raising an error might be reasonable.

jamescourtney replied:
For some examples, I have a few "nightmare struct" tests:

https://github.com/jamescourtney/FlatSharp/blob/2ec1312212655f5bf05e9678f3ec205aad3a410e/src/Tests/FlatSharpTests/SerializationTests/VectorSerializationTests.cs#L225
https://github.com/jamescourtney/FlatSharp/blob/2ec1312212655f5bf05e9678f3ec205aad3a410e/src/Tests/FlatSharpTests/SerializationTests/VectorSerializationTests.cs#L307

jamescourtney replied:
Finally, I think I might also suggest experimenting to see if this should be inlined or not. Does IL2CPP actually honor that attribute?

jamescourtney commented on src/FlatSharp.Runtime/IO/SpanWriterExtensions.cs:
```diff
@@ -36,6 +38,27 @@ public static class SpanWriterExtensions
         memory.Span.CopyTo(span.Slice(vectorStartOffset + sizeof(uint)));
     }
+    
+    public static void UnsafeWriteSpan<TSpanWriter, TElement>(
```
You actually don't need to pass all of these fields if you don't want. This is just the "default" serialization signature for FlatSharp's generated methods, but since this in effect a leaf node, feel free to throw other items in this signature as you need, such as item alignment, etc.

jamescourtney commented on src/FlatSharp.Runtime/IO/InputBufferExtensions.cs:
```diff
@@ -138,6 +139,23 @@ private static void ThrowInvalidVtableException()
+            // We need to construct a Span<TElement> from byte buffer that:
+            // 1. starts at correct offset for vector data
+            // 2. has a length based on *TElement* count not *byte* count
+            var byteSpanAtDataOffset = buffer.GetSpan().Slice(uoffset + sizeof(uint));
```
One note on `.GetSpan()`, which is that `ReadOnlyMemoryInputBuffer` will throw for this case. Since NativeArray is mutable, this seems appropriate to me, but just wanted to call it out.

vvuk commented on src/FlatSharp/TypeModel/Vectors/UnityNativeArrayVectorTypeModel.cs:
```diff
@@ -0,0 +1,97 @@
+                if (!buffer.IsPinned)
+                    throw new NotSupportedException(""Lazy parsing of a NativeArray requires a pinned buffer."");
+                var bufferSpan = {context.InputBufferVariableName}.UnsafeReadSpan<{context.InputBufferTypeName}, {this.ItemTypeModel.GetGlobalCompilableTypeName()}>({context.OffsetVariableName});
+                return Unity.Collections.LowLevel.Unsafe.NativeArrayUnsafeUtility.ConvertExistingDataToNativeArray<{this.ItemTypeModel.GetGlobalCompilableTypeName()}>(bufferSpan, Allocator.None);
```
Sadly, you'll have to also emit the ENABLE_COLLECTIONS_CHECKS #ifdef and the manual safety handle setting, despite Alloactor.None.  This is a super annoying wart with this API.   (This should crash if you try to read from one of these returned arrays in the editor or in development player builds currently.)

e.g.:

```
                   var foo = NativeArrayUnsafeUtility.ConvertExistingDataToNativeArray<>(...);

  #if ENABLE_UNITY_COLLECTIONS_CHECKS
                      NativeArrayUnsafeUtility.SetAtomicSafetyHandle(ref foo, AtomicSafetyHandle.GetTempUnsafePtrSliceHandle());
  #endif
                       return foo;
```

But I don't think TempUnsafePtrSliceHandle is correct. (It'll work fine for now.)  If we want to be _totally_ correct, we can allocate a safety handle that corresponds to the original buffer we're creating the array from, and use that.  That way if the input buffer goes away for some reason and someone kept a NativeArray around, they'll get a proper error.

jamescourtney reviewed (commented):
Looks good. Few more quibbles from me and some things I haven't thought about before, but just a bit of clarity on struct alignment and we can merge this.

Thanks!

jamescourtney commented:
@joncham  -- You may need to merge from main. One thing to keep in mind is that I've upgraded to target .net 7, which may cause Visual Studio to be unhappy. Unless you want to install the preview, just remove `net7.0` from the various .csproj elements to unblock yourself.

jamescourtney commented:
@joncham  -- are you waiting on me to review here? I did take a quick look at your latest changes, but saw that the alignment issue I mentioned before wasn't fixed, so I didn't add any new comments because I assumed you were not done. Let me know if I've missed something or you are waiting on input from me. Thanks!

joncham commented:
> @joncham -- are you waiting on me to review here? I did take a quick look at your latest changes, but saw that the alignment issue I mentioned before wasn't fixed, so I didn't add any new comments because I assumed you were not done. Let me know if I've missed something or you are waiting on input from me. Thanks!

@jamescourtney no, I am trying out the changes in some Unity scenarios to make sure things are working as I expect. I'll update PR to address your comments in coming days.

jamescourtney commented:
Hi @joncham  -- I'm going to take your breaking changes for `.IsPinned` and the `Memory<byte>` constructor and go ahead and commit those on their own. I'm eager to get Flatsharp v7 released and taking just these changes will make the rest of this not a breaking change.

joncham commented:
> Hi @joncham -- I'm going to take your breaking changes for `.IsPinned` and the `Memory<byte>` constructor and go ahead and commit those on their own. I'm eager to get Flatsharp v7 released and taking just these changes will make the rest of this not a breaking change.

Thanks @jamescourtney , I appreciate you grabbing those changes!

joncham commented:
Note, I pushed a rebased branch. I still didn't address all issues yet @jamescourtney but wanted to keep branch up to date.

jamescourtney commented:
> Note, I pushed a rebased branch. I still didn't address all issues yet @jamescourtney but wanted to keep branch up to date.

That sounds good. Sorry for any merge conflicts. I hope that you didn't have too difficult of a time. The branch is going to be mostly quiescent for awhile now. I'm starting to work on moving some test from reflection based execution to precompiled, but I won't be touching the main code for awhile. Thanks for your work here! Looking forward to publishing this when it's done :)

joncham commented:
@jamescourtney I am actively using this branch now in Unity, so I am comfortable enough merging if you are.

joncham commented:
Hey @jamescourtney just pinging again to get your thoughts.

jamescourtney commented:
> Hey @jamescourtney just pinging again to get your thoughts.

Hey! Sorry about this. I intentionally tried to unplug a bit during the holidays, but I'm (reluctantly) back to work now so will take a look at this soon.
